AAD/UserAADService.cs

```
@@ -0,0 +1,113 @@
+{
+    public class UserAADService
+    {
+        private readonly IConfiguration _configuration;
+        private readonly IConfidentialClientApplication _app;
+        private readonly IConfigurationSection _azureSettings;
+        private readonly SqlDatabase _database;
+        readonly MicrosoftIdentityConsentAndConditionalAccessHandler _consentHandler;
+
+        public UserAADService(IConfiguration configuration, SqlDatabase database, MicrosoftIdentityConsentAndConditionalAccessHandler consentHandler)
+        {
+            _consentHandler = consentHandler;
+
+            _database = database;
+
+            _configuration = configuration;
+
+            _azureSettings = _configuration.GetSection("AzureAd");
+
+            _app =
+                ConfidentialClientApplicationBuilder.Create(_azureSettings["ClientId"])
+                    .WithClientSecret(_azureSettings["ClientSecret"])
+                    .WithAuthority(AzureCloudInstance.AzurePublic, _azureSettings["TenantId"])
+                    .WithCacheOptions(CacheOptions.EnableSharedCacheOptions)
+                    .Build();
+        }
+
+        public async Task<string> GetAccessToken(AuthenticationState authState)
+        {
+            string accessToken = string.Empty;
+
+            //https://database.windows.net/.default
+            var scopes = new string[] { _azureSettings["Scopes"] };
+
+            try
+            {
+                var accountIdentifier = GetAccountIdentifier(authState);
+
+                IAccount account = await _app.GetAccountAsync(accountIdentifier);
+
+                AuthenticationResult authResult = await _app.AcquireTokenSilent(scopes, account).ExecuteAsync();
+                accessToken = authResult.AccessToken;
+            }
+            catch (MsalUiRequiredException)
+            {
+                _consentHandler.ChallengeUser(scopes);
+                return accessToken;
+            }
+
+            return accessToken;
```

This flow would return an empty string instead of throwing an exception.

Wouldn't we want to to pass on the information from the exception?

---

We want to suppress exception, empty token will be checked on the caller and empty summary list will be shown to user, by meanwhile user will be challenged in line 58. I preferred not to return exception to user at all

---

I disagree Alex, this is a bad pattern. Throw the exception with details like, new Exception("The call to Sql failed with the error <whatever>. <Some things they can look at to fix it, like "Ensure the user being logged-in exists as a user in the database " and all that>